metaflow/plugins/argo/argo_inner_flow.py

```diff
@@ -34,7 +68,139 @@ def __init__(
+        # Waiting for Argo workflow to trigger run is not optional.
+        # It is necessary to determine Flow name if not given as parameter
+        wait_to_trigger = 20  # wait time is 20 seconds
+        print(f"Attempting to trigger a run of Metaflow flow {self._flow_name}. Metaflow run id: {self._metaflow_run_id}, k8s namespace: {self._kubernetes_namespace}")
+        
+        start_time = time.time()
+        while wait_to_trigger > time.time() - start_time:
+            if self.has_triggered:
+                print(f"A run of Metaflow flow {self._flow_name} has started w/ Metaflow run id: {self._metaflow_run_id}")
+                break
+            elif self.finished:
+                # TODO: add specificity to exceptions (see AIP-6470)
+                raise Exception("Error - Unable to trigger run")
+            time.sleep(1)
+        else:
+            # TODO: add specificity to exceptions (see AIP-6470). MetaflowException (or a child exception in file) 
+            # or TimeoutError would be more specific.
+            raise Exception("Failed to begin running")
+
+        # optional wait for argo workflow to finish running
+        if self._wait:
+            print("\nNow we will wait for the flow to finish")
+
+            start_time = time.time()
+            loop_counter = 0
+            while self._wait_timeout * 60 > time.time() - start_time:
+                if self.finished:
+                    break
+                if loop_counter % 12 == 0:
+                    print(f"Time waited: {int((time.time() - start_time)/60)} minutes out of a possible {self._wait_timeout}")
+                loop_counter += 1
+                time.sleep(5)
+            else:
+                print("Run timed out.")
+                # TODO: add specificity to exceptions (see AIP-6470)
+                raise Exception("Wait Timeout")
+
+            success_statement = 'successfully!!!' if self.successful else 'unsuccessfully.'
+            print(f"\nTriggered run finished {success_statement}")
+
+        else:
+            print("\nNot waiting for run to finish")
+
+    @property
+    def _status(self):
+        if self._cached_status in ['Error', 'Failed', 'Succeeded']:
+            return self._cached_status
+        
+        workflow = self._argo_client.get_workflow(self._argo_run_id)
+        if workflow.get("status"):
+            self._cached_status = workflow["status"].get("phase")
+
+        return self._cached_status
+    
+    @property
+    def has_triggered(self):
+        if self._has_triggered:
+            return self._has_triggered
+        
+        elif self._status and self._status != 'Error':
+            self._has_triggered = True
+        
+        return self._has_triggered
+    
+    @property
+    def finished(self):
+        return self._status in ['Error', 'Failed', 'Succeeded']
+    
+    @property
+    def successful(self):
+        return self._status == 'Succeeded'
+    
+    @property
+    def failed_steps(self):
```

Comment:
style: please include the return Python hints for return types for every function

Reply:
done